rfc/0006-kubewarden-versioning.md

```diff
@@ -0,0 +1,203 @@
+|              |                                 |
+|:-------------|:--------------------------------|
+| Feature Name | Kubewarden versioning           |
+| Start Date   | Mar 3rd 2022                    |
+| Category     | Versioning                      |
+| RFC PR       | [fill this in after opening PR] |
+
+# Summary
+[summary]: #summary
+
+Kubewarden is composed of many core components that work together in
+order to build the final product. This is a proposal to harmonize
+Kubewarden versioning schema.
+
+# Motivation
+[motivation]: #motivation
+
+Having a way to specify what version of Kubewarden is being installed
+or executed would make it easier for the Kubewarden team to describe,
+market and support Kubewarden (as in user support).
+
+It would help users in understanding what version of Kubewarden they
+are about to install or execute.
+
+## Examples / User Stories
+[examples]: #examples
+
+> Users do not need to keep track of component versions and how they
+> work together. They just need to care about running a specific
+> Kubewarden version.
+
+> Users are able to easily understand that a new version of Kubewarden
+> is available, and that more than one component might be upgraded in
+> order to move to the new Kubewarden version.
+
+> A user has an issue with Kubewarden. Given its multiple component
+> nature, we would like to know what was the Kubewarden version that
+> got installed, and through what versions it was upgraded.
+
+# Detailed design
+[design]: #detailed-design
+
+## Many components, one experience
+
+Kubewarden core components, which are exposed to the user are:
+
+- Helm Charts
+- Kubewarden Controller
+- Policy Server
+- kwctl
+
+As a second dimension, there are also policies and their respective
+SDK's. This proposal is going to ignore this dimension in purpose,
+because Kubewarden runtime components (`policy-server` and `kwctl`)
+should be compatible with policies previously built, even with older
+language SDK's.
+
+### Public interfaces
+
+Our public interfaces with users are as follows:
+
+- Helm Charts
+  - `values.yaml`
+- Kubewarden Controller
+  - CLI args
+  - Custom Resource Definitions
+    - Versioned
+- Policy Server
+  - CLI args
+  - Configuration
+    - sigstore related files
+    - Usually handled by the Kubewarden Controller. Regardless we
+      should treat this as public interface, given some users might
+      run the Policy Server independent of the Kubewarden Controller,
+      and might configure it with the configuration file.
+- kwctl
+  - CLI args
+  - Configuration
+    - `sources.yaml`
+    - `docker.json`
+    - sigstore related files
+
+There is also the `waPC` contract between the `policy-evaluator`
+(consumed by the Policy Server and kwctl) and Policies and SDK's.
+
+## Kubewarden toplevel version
+
+This proposal envisions using [`CalVer`](https://calver.org/) to
+version the Kubewarden project.
+
```

Comment:
I don't see how calver would support modifiers, like -pre or -alpha etc.

Reply:
calver also doesn't show if a new version is a patch release, adds a new feature, or a breaking change.

Reply:
With calver we only release final versions. Is the "umbrella" toplevel Kubewarden version.

We can release -pre, -alpha of certain components if we desire.

Reply:
I like the psychological effect of Calver. It's cool to see a 2022.03 release and easy to understand. But when I think in how the Kubewarden is developed, I see us like a feature oriented project. Mostly of the time, we are not attached to date of things. So, this leds us to question like: `Should day versions be included in the toplevel map`? (from the `Unresolved question` section). Where we need to find a solution for when we have multiple global releases in the same time period.

It's not so uncommon we releasing features which require a "global release" (let's assume global release as a release where multiple components should be updated together) is a short period of time (less then a month). Therefore, I'm wondering if a simple version like v1, v2, v3 or even v1.0, v2.0 and so on... does not make more sense for our project. Which is also easy to understand and have a similar psychological effect of: "Wow! This project is advancing fast. We have 5 global releases within a month".

Reply:
Despite I believe calver is the optimal way to go for the umbrella version I think it has a problem: Kubewarden will stabilize, and the core components will not be updated so often (at some point, this will happen). When that happens, how do you feel about installing now a 21.03 version, today?

Is it stable? Or is it abandoned? This is the only thing that does not convince me about calver to be honest. I still think it's something we can overcome, the project can be active and posting blog posts, new policies, news...

Reply:
> Is it stable? Or is it abandoned? 
In that time , would we need to rebuild images because of CVEs, and then release a new calver? If yes, then, the calver version means "it is secure until this date, and potentially more if there hasn't been a new release version afterwards".

Reply:
Images are semver based.

Calver is the umbrella helm chart version. If we release today, 22.03.

If we fix a CVE on M07onday, we release 22.03.07.